### ISSUE TITLE
Flake: [Federation] Daemonset history CRUD verification failures in e2e

### ISSUE THREAD

**Issue by madhusudancs** [kubernetes/kubernetes#46925]
This started happening today evening. I don't think I have enough expertise to debug this test this quickly. This is blocking the queue right now. @csbell @marun @perotinus could one of you please take a look?

Could this be related to PR #45924 which merged around the same time we started seeing this flake?

/sig federation

/assign @csbell @marun @perotinus 

cc @shashidharatd @janetkuo

**Comment by madhusudancs** [comment]
Also cc @dchen1107 for the updates on this issue.

**Comment by csbell** [comment]
This appears to be related to recent changes introduced by Daemonset history PR #45924. It is flaky based on CI and pull outputs.

@madhusudancs noted: Despite the output in CI which attributes test failures to either configmap, secrets or daemonsets, the errors are in fact all daemonsets (this is due to a mismatch between the ginkgo selection criteria and how the CRUD-based testing is done in the sync controller).

@janetkuo knows about this and @madhusudancs is making some underlying federation changes to help in debugging.

**Comment by madhusudancs** [comment]
While setting up federation e2e tests to extract additional information to help @janetkuo debug this problem, I found that the object that we send to clusters and the object that we receive have different labels: Here is the object that we send to clusters:

```yaml
&v1beta1.DaemonSet{
    TypeMeta: v1.TypeMeta{
        Kind: "",
        APIVersion: "",
    },
    ObjectMeta: v1.ObjectMeta{
        Name: "test-daemonset-zxw0m",
        GenerateName: "test-daemonset-",
        Namespace: "e2e-tests-federated-types-q50mx",
        SelfLink: "/apis/extensions/v1beta1/namespaces/e2e-tests-federated-types-q50mx/daemonsets/test-daemonset-zxw0m",
        UID: "5174e9ba-4a47-11e7-a2cf-0a580ab60308",
        ResourceVersion: "733",
        Generation: 1,
        CreationTimestamp: v1.Time{
            Time: time.Time{
                sec: 63632302382,
                nsec: 0,
                loc: (*time.Location)(0x5afa540),
            },
        },
        DeletionTimestamp: (*v1.Time)(nil),
        DeletionGracePeriodSeconds: (*int64)(nil),
        Labels: map[string]string{
            "app": "test-daemonset",
        },
        Annotations: map[string]string{
            "federation.kubernetes.io/test-federation-crud-update": "updated",
        },
        OwnerReferences: []v1.OwnerReference(nil),
        Initializers: (*v1.Initializers)(nil),
        Finalizers: []string{
            "federation.kubernetes.io/delete-from-underlying-clusters",
            "orphan",
        },
        ClusterName: "",
    },
    Spec: v1beta1.DaemonSetSpec{
        Selector: (*v1.LabelSelector)(0xc420113500),
        Template: v1.PodTemplateSpec{
            ObjectMeta: v1.ObjectMeta{
                Name: "",
                GenerateName: "",
                Namespace: "",
                SelfLink: "",
                UID: "",
                ResourceVersion: "",
                Generation: 0,
                CreationTimestamp: v1.Time{
                    Time: time.Time{
                        sec: 0,
                        nsec: 0,
                        loc: (*time.Location)(nil),
                    },
                },
                DeletionTimestamp: (*v1.Time)(nil),
                DeletionGracePeriodSeconds: (*int64)(nil),
                Labels: map[string]string{
                    "name": "test-pod",
                },
                Annotations: map[string]string(nil),
                OwnerReferences: []v1.OwnerReference(nil),
                Initializers: (*v1.Initializers)(nil),
                Finalizers: []string(nil),
                ClusterName: "",
            },
            Spec: v1.PodSpec{
                Volumes: []v1.Volume(nil),
                InitContainers: []v1.Container(nil),
                Containers: []v1.Container{
                    v1.Container{
                        Name: "test-daemonset",
                        Image: "images/test-daemonset",
                        Command: []string(nil),
                        Args: []string(nil),
                        WorkingDir: "",
                        Ports: []v1.ContainerPort{
                            v1.ContainerPort{
                                Name: "",
                                HostPort: 0,
                                ContainerPort: 9376,
                                Protocol: "TCP",
                                HostIP: "",
                            },
                        },
                        EnvFrom: []v1.EnvFromSource(nil),
                        Env: []v1.EnvVar(nil),
                        Resources: v1.ResourceRequirements{
                            Limits: v1.ResourceList(nil),
                            Requests: v1.ResourceList(nil),
                        },
                        VolumeMounts: []v1.VolumeMount(nil),
                        LivenessProbe: (*v1.Probe)(nil),
                        ReadinessProbe: (*v1.Probe)(nil),
                        Lifecycle: (*v1.Lifecycle)(nil),
                        TerminationMessagePath: "/dev/termination-log",
                        TerminationMessagePolicy: "File",
                        ImagePullPolicy: "Always",
                        SecurityContext: (*v1.SecurityContext)(nil),
                        Stdin: false,
                        StdinOnce: false,
                        TTY: false,
                    },
                },
                RestartPolicy: "Always",
                TerminationGracePeriodSeconds: (*int64)(0xc420191390),
                ActiveDeadlineSeconds: (*int64)(nil),
                DNSPolicy: "ClusterFirst",
                NodeSelector: map[string]string(nil),
                ServiceAccountName: "",
                DeprecatedServiceAccount: "",
                AutomountServiceAccountToken: (*bool)(nil),
                NodeName: "",
                HostNetwork: false,
                HostPID: false,
                HostIPC: false,
                SecurityContext: (*v1.PodSecurityContext)(0xc420485540),
                ImagePullSecrets: []v1.LocalObjectReference(nil),
                Hostname: "",
                Subdomain: "",
                Affinity: (*v1.Affinity)(nil),
                SchedulerName: "default-scheduler",
                Tolerations: []v1.Toleration(nil),
                HostAliases: []v1.HostAlias(nil),
            },
        },
        UpdateStrategy: v1beta1.DaemonSetUpdateStrategy{
            Type: "OnDelete",
            RollingUpdate: (*v1beta1.RollingUpdateDaemonSet)(nil),
        },
        MinReadySeconds: 0,
        TemplateGeneration: 1,
        RevisionHistoryLimit: (*int32)(0xc4201913c0),
    },
    Status: v1beta1.DaemonSetStatus{
        CurrentNumberScheduled: 0,
        NumberMisscheduled: 0,
        DesiredNumberScheduled: 0,
        NumberReady: 0,
        ObservedGeneration: 0,
        UpdatedNumberScheduled: 0,
        NumberAvailable: 0,
        NumberUnavailable: 0,
        CollisionCount: (*int64)(nil),
    },
}
```

Object returned by a cluster:

```yaml
&v1beta1.DaemonSet{
    TypeMeta: v1.TypeMeta{
        Kind: "",
        APIVersion: "",
    },
    ObjectMeta: v1.ObjectMeta{
        Name: "test-daemonset-zxw0m",
        GenerateName: "",
        Namespace: "e2e-tests-federated-types-q50mx",
        SelfLink: "/apis/extensions/v1beta1/namespaces/e2e-tests-federated-types-q50mx/daemonsets/test-daemonset-zxw0m",
        UID: "517827ad-4a47-11e7-9fbe-42010af00002",
        ResourceVersion: "20188",
        Generation: 1,
        CreationTimestamp: v1.Time{
            Time: time.Time{
                sec: 63632302382,
                nsec: 0,
                loc: (*time.Location)(0x5afa540),
            },
        },
        DeletionTimestamp: (*v1.Time)(nil),
        DeletionGracePeriodSeconds: (*int64)(nil),
        Labels: map[string]string{
            "daemonset-controller-hash": "3497893742",
            "app": "test-daemonset",
        },
        Annotations: map[string]string{
            "federation.kubernetes.io/test-federation-crud-update": "updated",
        },
        OwnerReferences: []v1.OwnerReference(nil),
        Initializers: (*v1.Initializers)(nil),
        Finalizers: []string(nil),
        ClusterName: "",
    },
    Spec: v1beta1.DaemonSetSpec{
        Selector: (*v1.LabelSelector)(0xc421421da0),
        Template: v1.PodTemplateSpec{
            ObjectMeta: v1.ObjectMeta{
                Name: "",
                GenerateName: "",
                Namespace: "",
                SelfLink: "",
                UID: "",
                ResourceVersion: "",
                Generation: 0,
                CreationTimestamp: v1.Time{
                    Time: time.Time{
                        sec: 0,
                        nsec: 0,
                        loc: (*time.Location)(nil),
                    },
                },
                DeletionTimestamp: (*v1.Time)(nil),
                DeletionGracePeriodSeconds: (*int64)(nil),
                Labels: map[string]string{
                    "name": "test-pod",
                },
                Annotations: map[string]string(nil),
                OwnerReferences: []v1.OwnerReference(nil),
                Initializers: (*v1.Initializers)(nil),
                Finalizers: []string(nil),
                ClusterName: "",
            },
            Spec: v1.PodSpec{
                Volumes: []v1.Volume(nil),
                InitContainers: []v1.Container(nil),
                Containers: []v1.Container{
                    v1.Container{
                        Name: "test-daemonset",
                        Image: "images/test-daemonset",
                        Command: []string(nil),
                        Args: []string(nil),
                        WorkingDir: "",
                        Ports: []v1.ContainerPort{
                            v1.ContainerPort{
                                Name: "",
                                HostPort: 0,
                                ContainerPort: 9376,
                                Protocol: "TCP",
                                HostIP: "",
                            },
                        },
                        EnvFrom: []v1.EnvFromSource(nil),
                        Env: []v1.EnvVar(nil),
                        Resources: v1.ResourceRequirements{
                            Limits: v1.ResourceList(nil),
                            Requests: v1.ResourceList(nil),
                        },
                        VolumeMounts: []v1.VolumeMount(nil),
                        LivenessProbe: (*v1.Probe)(nil),
                        ReadinessProbe: (*v1.Probe)(nil),
                        Lifecycle: (*v1.Lifecycle)(nil),
                        TerminationMessagePath: "/dev/termination-log",
                        TerminationMessagePolicy: "File",
                        ImagePullPolicy: "Always",
                        SecurityContext: (*v1.SecurityContext)(nil),
                        Stdin: false,
                        StdinOnce: false,
                        TTY: false,
                    },
                },
                RestartPolicy: "Always",
                TerminationGracePeriodSeconds: (*int64)(0xc420327600),
                ActiveDeadlineSeconds: (*int64)(nil),
                DNSPolicy: "ClusterFirst",
                NodeSelector: map[string]string(nil),
                ServiceAccountName: "",
                DeprecatedServiceAccount: "",
                AutomountServiceAccountToken: (*bool)(nil),
                NodeName: "",
                HostNetwork: false,
                HostPID: false,
                HostIPC: false,
                SecurityContext: (*v1.PodSecurityContext)(0xc4202c83c0),
                ImagePullSecrets: []v1.LocalObjectReference(nil),
                Hostname: "",
                Subdomain: "",
                Affinity: (*v1.Affinity)(nil),
                SchedulerName: "default-scheduler",
                Tolerations: []v1.Toleration(nil),
                HostAliases: []v1.HostAlias(nil),
            },
        },
        UpdateStrategy: v1beta1.DaemonSetUpdateStrategy{
            Type: "OnDelete",
            RollingUpdate: (*v1beta1.RollingUpdateDaemonSet)(nil),
        },
        MinReadySeconds: 0,
        TemplateGeneration: 1,
        RevisionHistoryLimit: (*int32)(0xc420327620),
    },
    Status: v1beta1.DaemonSetStatus{
        CurrentNumberScheduled: 4,
        NumberMisscheduled: 0,
        DesiredNumberScheduled: 4,
        NumberReady: 0,
        ObservedGeneration: 1,
        UpdatedNumberScheduled: 4,
        NumberAvailable: 0,
        NumberUnavailable: 4,
        CollisionCount: (*int64)(nil),
    },
}
```

**Comment by madhusudancs** [comment]
@janetkuo and I briefly chatted about this problem. We will discuss this further tomorrow. The summary is, underlying clusters modifying the spec or the metadata is a problem in general. We will see where else we can store the pod template hash in the DaemonSet object.

**Comment by madhusudancs** [comment]
@janetkuo and I discussed about this problem and we have resolved to apply a short term bandaid for this problem in federation. The way we are going to do this is by ignoring the `daemonset-controller-hash` label on the DaemonSet objects if they exist, while comparing the Federated DaemonSet and cluster DaemonSet objects.

@csbell points out that this is not tenable for version skew. I think this shouldn't affect version skew if we do nothing if the label doesn't exist.

@csbell @nikhiljindal @marun @shashidharatd @perotinus let me know if you have any comments about this. I am going to send a PR soon.
